ui/page/governance/agenda_vote_modal.go

```diff
@@ -0,0 +1,307 @@
+package governance
+
+import (
+	"sort"
+
+	"gioui.org/font/gofont"
+	"gioui.org/layout"
+	"gioui.org/text"
+	"gioui.org/widget"
+	"gioui.org/widget/material"
+
+	"github.com/planetdecred/dcrlibwallet"
+	"github.com/planetdecred/godcr/ui/decredmaterial"
+	"github.com/planetdecred/godcr/ui/load"
+	"github.com/planetdecred/godcr/ui/page/components"
+	"github.com/planetdecred/godcr/ui/values"
+)
+
+type agendaVoteModal struct {
+	*load.Load
+	modal decredmaterial.Modal
+
+	LiveTickets []*dcrlibwallet.Transaction
+
+	agenda    *dcrlibwallet.Agenda
+	isVoting  bool
+	loadCount int
+
+	consensusPage *ConsensusPage
+
+	walletSelector    *WalletSelector
+	ticketSelector    *ticketSelector
+	spendingPassword  decredmaterial.Editor
+	materialLoader    material.LoaderStyle
+	items             map[string]string //[key]str-key
+	itemKeys          []string
+	defaultValue      string // str-key
+	initialValue      string
+	currentValue      string
+	optionsRadioGroup *widget.Enum
+	voteBtn           decredmaterial.Button
+	cancelBtn         decredmaterial.Button
+}
+
+func newAgendaVoteModal(l *load.Load, agenda *dcrlibwallet.Agenda, consensusPage *ConsensusPage) *agendaVoteModal {
+	avm := &agendaVoteModal{
+		Load:              l,
+		modal:             *l.Theme.ModalFloatTitle(),
+		agenda:            agenda,
+		consensusPage:     consensusPage,
+		materialLoader:    material.Loader(material.NewTheme(gofont.Collection())),
+		optionsRadioGroup: new(widget.Enum),
+		spendingPassword:  l.Theme.EditorPassword(new(widget.Editor), "Spending password"),
+		voteBtn:           l.Theme.Button("Update Preference"),
+		cancelBtn:         l.Theme.OutlineButton("Cancel"),
+	}
+
+	avm.voteBtn.Background = l.Theme.Color.Gray3
+	avm.voteBtn.Color = l.Theme.Color.Surface
+
+	avm.walletSelector = NewWalletSelector(l).
+		Title("Voting wallet").
+		WalletSelected(func(w *dcrlibwallet.Wallet) {
+			avm.loadCount = 0
+
+			avm.FetchUnspentUnexpiredTickets(w.ID)
+			avm.RefreshWindow()
+			avm.loadCount++
+
+			// update agenda options prefrence to that of the selected wallet
+			consensusItems := components.LoadAgendas(avm.Load, w, false)
+			for _, consensusItem := range consensusItems {
+				if consensusItem.Agenda.AgendaID == agenda.AgendaID {
+					ArrVoteOptions := make(map[string]string)
+					for i := range consensusItem.Agenda.Choices {
+						ArrVoteOptions[agenda.Choices[i].Id] = consensusItem.Agenda.Choices[i].Id
+					}
+
+					// sort keys to keep order when refreshed
+					sortedKeys := make([]string, 0)
+					for k := range ArrVoteOptions {
+						sortedKeys = append(sortedKeys, k)
+					}
+
+					sort.Slice(sortedKeys, func(i, j int) bool {
+						var stringI string = sortedKeys[i]
+						var stringJ string = sortedKeys[j]
+						if stringI == stringJ {
+							return stringI < stringJ
+						}
+						return stringI < stringJ
+					})
```

Comment:
If you use an array instead of a map, you shouldn't need to sort. Also, the key is the same as the value, even though read from different objects; because `agenda` is the same as `consensusItem.Agenda`.
`ArrVoteOptions := make(map[string]string)` should be `voteChoices := make([]string, len(consensusItem.Agenda.Choices))`